File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraysWithKDistinct = function(nums, k) {
  return atMost(nums, k) - atMost(nums, k - 1);
};

const atMost = (nums, k) => {
  const count = {};
  let [l, r, ans] = [0, 0, 0];
  
  while (r < nums.length) {
    if (!(nums[r] in count)) count[nums[r]] = 0;
    if (!count[nums[r]]++) k--;
    
    while (k < 0) {
      if (!--count[nums[l++]]) k++;
    }
    
    ans += r - l + 1;
    r++;
  }
  
  return ans;
}

File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.py
class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        def atMost(nums, k):
            count = Counter()
            j = ans = 0
            for i, num in enumerate(nums):
                k -= count[num] == 0
                count[num] += 1
                while k < 0:
                    count[nums[j]] -= 1
                    k += count[nums[j]] == 0                    
                    j += 1
                ans += i - j + 1
            return ans
          
          
        return atMost(nums, k) - atMost(nums, k - 1)
                    